solr/core/src/java/org/apache/solr/search/CaffeineCache.java

```diff
@@ -269,7 +289,38 @@ public V computeIfAbsent(K key, IOFunction<? super K, ? extends V> mappingFuncti
           });
     } catch (UncheckedIOException e) {
       throw e.getCause();
+    } catch (RecursionException ignored) {
+      // recursive invocation detected; fallback to get-then-put
+    }
+
```

Comment:
At this line, a comment saying "Recursive update detected; fall back on get-then-put" would clarify the overall flow of this complicated method.